onnxruntime/core/framework/allocatormgr.cc

```diff
@@ -54,4 +61,38 @@ AllocatorPtr CreateAllocator(const AllocatorCreationInfo& info) {
   return AllocatorPtr(std::move(device_allocator));
 }
 
+
+// Update allocator in the provider if already present; ignore if not.
+void AllocatorManager::ReplaceAllocator(AllocatorPtr allocator) {
+  const auto& info = allocator->Info();
+  auto ite = mem_info_set_.find(info);
+  if (ite != mem_info_set_.end()) {
+    const int key = MakeKey(info.id, info.mem_type);
+    allocators_[key] = allocator;
+  }
+}
+
+void AllocatorManager::InsertAllocator(AllocatorPtr allocator) {
+  const OrtMemoryInfo& info = allocator->Info();
+  auto ite = mem_info_set_.find(info);
+  if (ite != mem_info_set_.end()) {
+    LOGS_DEFAULT(WARNING) << "duplicated allocator: " << info.ToString();
+    return;
```

Comment:
Maybe better to return Status of failure, and let the outside logic to handle failure properly. Warning usually is ignored.